packages/cli/src/commands/onboard/discovery/providers/github/graphql.ts

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { graphql } from '@octokit/graphql';
+
+/**
+ * Assists in repeatedly executing a query with a paged response.
+ *
+ * Requires that the query accepts a $cursor variable.
+ *
+ * @param client - The octokit client
+ * @param query - The query to execute
+ * @param org - The slug of the org to read
+ * @param connection - A function that, given the response, picks out the actual
+ *                   Connection object that's being iterated
+ * @param transformer - A function that, given one of the nodes in the
+ *               Connection, returns the model mapped form of it
+ * @param variables - The variable values that the query needs, minus the cursor
+ */
+export async function queryWithPaging<
+  GraphqlType,
+  OutputType,
+  Variables extends {},
+  Response = QueryResponse,
+>(
+  client: typeof graphql,
+  query: string,
+  org: string,
+  connection: (response: Response) => Connection<GraphqlType> | undefined,
+  transformer: (
+    item: GraphqlType,
+    ctx: TransformerContext,
+  ) => Promise<OutputType | undefined>,
+  variables: Variables,
+): Promise<OutputType[]> {
+  const result: OutputType[] = [];
+
+  let cursor: string | undefined = undefined;
+  for (let j = 0; j < 1000 /* just for sanity */; ++j) {
```

Comment:
If we hit `1000`, this will fail silently with missing information.
I'd lean towards `while (hasNextPage)` instead, because at least if it hangs, we'll definitely get a bug report 👍

Reply:
https://github.com/backstage/backstage/pull/17624/commits/73965ce76ab59987ec8c56c90200f6504c769b04